crates/sim2h/src/connection_mgr.rs

```diff
@@ -0,0 +1,319 @@
+use crate::*;
+use std::sync::{Arc, Weak};
+
+/// incoming messages from websockets
+#[derive(Debug)]
+pub enum ConMgrEvent {
+    Disconnect(Lib3hUri, Option<Sim2hError>),
+    ReceiveData(Lib3hUri, WsFrame),
+    ConnectionCount(usize),
+}
+
+/// messages for controlling the connection manager
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+enum ConMgrCommand {
+    Connect(Lib3hUri, TcpWss),
+    SendData(Lib3hUri, WsFrame),
+    Disconnect(Lib3hUri),
+}
+
+type EvtSend = tokio::sync::mpsc::UnboundedSender<ConMgrEvent>;
+type EvtRecv = tokio::sync::mpsc::UnboundedReceiver<ConMgrEvent>;
+type CmdSend = tokio::sync::mpsc::UnboundedSender<ConMgrCommand>;
+type CmdRecv = tokio::sync::mpsc::UnboundedReceiver<ConMgrCommand>;
+
+pub type ConnectionMgrEventRecv = EvtRecv;
+
+/// internal websocket polling loop
+async fn wss_task(uri: Lib3hUri, mut wss: TcpWss, evt_send: EvtSend, mut cmd_recv: CmdRecv) {
+    let mut frame = None;
+
+    // TODO - this should be done with tokio tcp streams && selecting
+    //        for now, we're just pausing when no work happens
+
+    loop {
+        let mut did_work = false;
+
+        // first, process a batch of control commands
+        for _ in 0..100 {
+            match cmd_recv.try_recv() {
+                Ok(cmd) => {
+                    did_work = true;
+                    match cmd {
+                        ConMgrCommand::SendData(_uri, frame) => {
+                            if let Err(e) = wss.write(frame) {
```

Comment:
anyway to get an async write in here? Just wondering if it is possible

Reply:
Would very much love to switch to [tokio-tungstenite](https://crates.io/crates/tokio-tungstenite) - but alas, the currently published version isn't compatible with the version of tokio we're using - They're working on it.

The other [tokio websocket library](https://crates.io/crates/websocket) says it's not being maintained any more and to use tokio-tungstenite.

For now, at least, the read / write sockets we're using are non-blocking so it won't block the thread if they're not ready to return data immediately.

Reply:
Awesome!